Add option to disable synchronization traffic for RGWs

polyedre opened the pull request:
**Description of your changes:**

Some users want to deploy two CephObjectStores for a single Zone. The first configures RGWs to process the synchronization of the data, while the second CephObjectStore configures the client RGWs.

Currently, this can be implemented by setting the RGW option 'rgw_run_sync_thread' in the 'rook-config-override' ConfigMap, though it is not really user friendly.

Ref: https://docs.ceph.com/en/latest/radosgw/config-ref/#confval-rgw_run_sync_thread

This commit adds a new option in the CephObjectStore CRD as defined in issue https://github.com/rook/rook/issues/12272. The new option 'disableMultisiteSyncTraffic' determine whether the operator should disable the multisite sync threads for the RGWs.

If the option is set to 'false', or if the option is not specified, the operator does nothing. This ensures that the multisite sync threads will not be enabled for users that disabled explicitely the multisite sync threads either manually or with the 'rook-config-override' ConfigMap.

**Tests**

I manually tested the result of the `make build` by deploying the helm charts. I validated that when the option is not configured in the CephObjectStore, the option `rgw_run_sync_thread` is not configured in the RGW, and that when `disableMultisiteSyncTraffic` is set to `true`, the same option is set to `false` in the RGW.

Please note that setting `disableMultisiteSyncTraffic` to `false` will result in the option beeing left untouched in ceph. 

**Which issue is resolved by this Pull Request:**
Resolves #12272

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -1433,6 +1433,12 @@ type GatewaySpec struct {
 
+	// Whether to disable the Multisite synchronization traffic for the rgw daemon. If not set, the
+	// multisite synchronization traffic will be enabled.
+	// +nullable
```
Only pointers are nullable. Remove this.

BlaineEXE replied:
Thanks for handling line wrapping!

BlaineEXE replied:
Still marked nullable.

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -1433,6 +1433,12 @@ type GatewaySpec struct {
 	// +optional
 	Placement Placement `json:"placement,omitempty"`
 
+	// Whether to disable the Multisite synchronization traffic for the rgw daemon. If not set, the
+	// multisite synchronization traffic will be enabled.
+	// +nullable
+	// +optional
+	DisableMultisiteSyncTraffic bool `json:"disableMultisiteSyncTraffic,omitempty"`
```
Per Casey and Ali's comments, let's rework this a little to be extra clear for users that this only affects the send/transmit portion of multisite sync.

```suggestion
	// Prevent this object store's gateways from transmitting multisite replication data. Note that this this value does not affect whether gateways receive multisite replication traffic. If false or unset, this object store's gateways will be able to transmit multisite replication data. 
	// +optional
	DisableMultisiteSyncTraffic bool `json:"disableMultisiteSyncTraffic,omitempty"`
```

BlaineEXE commented:
> I had a look at the unit tests and I am not sure how to implement one for this specific feature. Should I mock the `monStore` and validate that the RGW option is set to false if `disableMultisiteSyncTraffic` is set to true ? It seems there is a `MockExecutor`, but that `monStore.Set` uses the `RemoteExecutor` which do not have a mock.

https://github.com/rook/rook/blob/911ce938327fef0cbae38bdcec1b2d3d3237ae7e/pkg/operator/ceph/object/rgw_test.go#L40-L72

It should be fairly straightforward to extend the test linked above to test this PR as well. Please encapsulate the current test into subtest via `t.Run(...)`, and then create a second subtest for testing enable/disable of multisite replication. You can simply copy-paste the setup into the new subtest unless you want to factor out shared code. IMO, that's not strictly necessary here; correctness is more important.

The new test will have to modify the mock executor to allow `mon config set...` commands, and it should verify that the right config key and value are used. I often use `t.Log()` to see what the mock executor is receiving from Rook to help know what to set up. 

The new test should test the `set false` case first, followed by the `set true` case to ensure that sync enable is re-allowed if the value is removed from the CRD after first being set.

BlaineEXE commented:
https://github.com/rook/rook/blob/b2babf01d279b2002403623cd3482cdf6d2de73f/pkg/apis/ceph.rook.io/v1/types.go#L1761-L1769

Because we are making a distinction here that this only controlls multisite data transmission, it would be good to add to the documentation linked: "If a CephObjectStore endpoint is omitted from this list, that object store's gateways will not receive multisite replication data (see CephObjectStore.spec.gateway.disableMultisiteDataTransmission)."

polyedre commented on pkg/operator/ceph/object/rgw.go:
```diff
@@ -145,6 +146,16 @@ func (c *clusterConfig) startRGWPods(realmName, zoneGroupName, zoneName string)
 			}
 		}
 
+		// Disable multisite sync traffic
+		if c.store.Spec.Gateway.DisableMultisiteSyncTraffic {
+			monStore := cephconfig.GetMonStore(c.context, c.clusterInfo)
+			who := generateCephXUser(rgwConfig.ResourceName)
+			err := monStore.Set(who, "rgw_run_sync_thread", "false")
+			if err != nil {
+				return errors.Wrapf(err, "failed to set rgw_run_sync_thread to false on %q", who)
+			}
+		}
```
Thanks for pointing this out! I changed the option from a bool to a pointer so that I can detect whether the option has been specified by the users. To avoid any override of user-defined values via ceph commands applied from the toolbox, I propose to do nothing if the option is not set.

polyedre commented:
I've integrated the comments from the code review and implemented the unit tests.

I changed the type of `DisableMultisiteSyncTraffic` from a `bool` to a `*bool` to be able to detect whether the users specified the option manually.This should ensure the application of the gateways configuration is idempotent when the option is enabled while avoiding any override of the option for users that configure the option themselves.

I added the new work as new commits. What is the policy of Rook regarding the git history? Should I rewrite the history to have only 1 stashed commit? Or should commits that integrate comments from the code review be fixups of previous commits?

BlaineEXE commented on pkg/operator/ceph/object/rgw.go:
```diff
@@ -145,6 +146,20 @@ func (c *clusterConfig) startRGWPods(realmName, zoneGroupName, zoneName string)
 		}
 
+		// Enable or disable multisite sync traffic
+		if c.store.Spec.Gateway.DisableMultisiteSyncTraffic != nil {
+			doMultisiteSync := "true"
+			if *c.store.Spec.Gateway.DisableMultisiteSyncTraffic {
+				doMultisiteSync = "false"
+			}
+			monStore := cephconfig.GetMonStore(c.context, c.clusterInfo)
+			who := generateCephXUser(rgwConfig.ResourceName)
+			err := monStore.Set(who, "rgw_run_sync_thread", doMultisiteSync)
+			if err != nil {
+				return errors.Wrapf(err, "failed to set rgw_run_sync_thread to %q on %q", doMultisiteSync, who)
+			}
+		}
```
This can still result in lost configurations and is not a proper implementation of an idempotent desired-state system. If a user sets `disableMultisiteDataTransmission: true` and then removes the value entirely, the config will not be reset to its default value. 

This is a bug caused by premature optimization stemming from an attempt to reduce the number of "set" calls issued to Ceph by Rook. Tthis is a minor cost, and because Rook is a management engine, performance is not a primary concern.

Please implement as I previously suggested, remove the pointer on the bool value, and remove the nullable option.

BlaineEXE commented on pkg/operator/ceph/object/rgw_test.go:
```diff
@@ -65,10 +79,52 @@ func TestStartRGW(t *testing.T) {
+		validateStart(ctx, t, c, clientset)
+	})
+
+	t.Run("Multisite sync traffic not set", func(t *testing.T) {
+		store.ObjectMeta.Name = "client-rgw-option-not-set"
+		err := c.startRGWPods(store.Name, store.Name, store.Name)
+		assert.Nil(t, err)
+
+		assert.NotContains(t, appliedRgwConfigurations, "rgw_run_sync_thread")
+	})
```
I requested a unit test that tests this:

> The new test should test the set false case first, followed by the set true case to ensure that sync enable is re-allowed if the value is removed from the CRD after first being set.

The critical part here is "the value is removed from the CRD after first being set". This would have caught the optimization bug that I flagged before.

polyedre commented:
I'm sorry I didn't implement the changes requested in the first code review. By converting the option into a pointer to a boolean, I was thinking not of avoiding the call to the 'Set' method, but of avoiding a breakable behavior for users who modified rgw_run_sync_thread manually. For these users, updating Rook after this PR could rewrite the variable they set.

I now understand that idempotency is more important in a desired-state system context. So I've gone back to the requested changes.

polyedre commented:
Hi! I think this PR address the requested changes. Can someone review this please? @travisn @BlaineEXE

Thanks!

parth-gr commented:
Can you please squash all the commits into a single commit?

We generally keep 1 commit 1 PR if the changes are for the same feature

polyedre commented on pkg/operator/ceph/object/rgw.go:
```diff
@@ -145,6 +146,17 @@ func (c *clusterConfig) startRGWPods(realmName, zoneGroupName, zoneName string)
+		}
+		monStore := cephconfig.GetMonStore(c.context, c.clusterInfo)
+		who := generateCephXUser(rgwConfig.ResourceName)
+		err = monStore.Set(who, "rgw_run_sync_thread", doMultisiteSync)
```
I found some time to work on this again: what you propose is way cleaner! The tests still pass but I'll need to check the behavior again against a cluster to be sure.

I kept using `monStore.Set()`. I'll look at using `monStore.SetAll()` when I find some time.

BlaineEXE commented:
> I'm sorry I didn't implement the changes requested in the first code review. By converting the option into a pointer to a boolean, I was thinking not of avoiding the call to the 'Set' method, but of avoiding a breakable behavior for users who modified rgw_run_sync_thread manually. For these users, updating Rook after this PR could rewrite the variable they set.
> 
> I now understand that idempotency is more important in a desired-state system context. So I've gone back to the requested changes.

That is a good point that this feature could rewrite a user-modified variable. We should mention this in release notes as a potential breaking change for those users.

Since we are so close to the Rook v1.12 release, let's target getting this into the new release where this kind of change behavior is more likely to be noticed by users. :)

BlaineEXE commented:
This looks good to me from code/doc perspective. @alimaredia do you think you might have time to test out the changes for any corner cases you think could be easy to miss?

polyedre commented:
Thanks for taking some time to review this!

When testing the replication with the two CephObjectStore scenario, we discovered that the replication was broken when a gateway that do run sync threads is added to the Zone endpoints.

If this is the expected behavior of Ceph's replication, I think we need to block the CephObjectStore to add its own endpoint to the Zone when DisableMultisiteSyncTraffic is set to `true`.

I had a look at the code. It seems that the method `setMultisite` in `pkg/operator/ceph/object/objectstore.go` should not be called when `DisableMultisiteSyncTraffic` is set to `true`.

If stopping calling the line https://github.com/rook/rook/blob/647a4e3534a41015aae9bc3ed21e12b10a9f9192/pkg/operator/ceph/object/controller.go#L447 is okay to you, I'll update this PR to take this into account. We may need a new unit test too.
